lib/logproto/logproto-auto-server.c

```diff
@@ -0,0 +1,151 @@
+ *
+ */
+#include "logproto-auto-server.h"
+#include "logproto-text-server.h"
+#include "logproto-framed-server.h"
+#include "messages.h"
+
+typedef struct _LogProtoAutoServer
+{
+  LogProtoServer super;
+
+  /* the actual LogProto instance that we run after auto-detecting the protocol */
+  LogProtoServer *proto_impl;
+} LogProtoAutoServer;
+
+static LogProtoServer *
+_construct_detected_proto(LogProtoAutoServer *self, const gchar *detect_buffer, gsize detect_buffer_len)
+{
+  if (detect_buffer_len != 1)
+    goto error;
+
+  if (detect_buffer[0] == '<')
+    {
+      msg_debug("Auto-detected non-transparent-framing on RFC6587 connection, using simple text transport",
+                evt_tag_int("fd", self->super.transport->fd));
+      return log_proto_text_server_new(self->super.transport, self->super.options);
+    }
+  else if (g_ascii_isdigit(detect_buffer[0]))
+    {
+      msg_debug("Auto-detected octet-counted-framing on RFC6587 connection, using framed transport",
+                evt_tag_int("fd", self->super.transport->fd));
+      return log_proto_framed_server_new(self->super.transport, self->super.options);
+    }
+
+error:
+  msg_debug("Unable to detect framing on RFC6587 connection",
+            evt_tag_int("fd", self->super.transport->fd),
+            evt_tag_mem("detect_buffer", detect_buffer, detect_buffer_len));
+  return NULL;
+}
+
+static LogProtoPrepareAction
+log_proto_auto_server_prepare(LogProtoServer *s, GIOCondition *cond, gint *timeout G_GNUC_UNUSED)
+{
+  LogProtoAutoServer *self = (LogProtoAutoServer *) s;
+
+  if (self->proto_impl)
+    return log_proto_server_prepare(self->proto_impl, cond, timeout);
+
+  *cond = self->super.transport->cond;
+  if (*cond == 0)
+    *cond = G_IO_IN;
+
+  return LPPA_POLL_IO;
+}
+
+static LogProtoStatus
+log_proto_auto_server_fetch(LogProtoServer *s, const guchar **msg, gsize *msg_len, gboolean *may_read,
+                            LogTransportAuxData *aux, Bookmark *bookmark)
+{
+  LogProtoAutoServer *self = (LogProtoAutoServer *) s;
+
+  if (self->proto_impl)
+    return log_proto_server_fetch(self->proto_impl, msg, msg_len, may_read, aux, bookmark);
+
+  g_assert_not_reached();
+}
+
+static gboolean
+log_proto_auto_handshake_in_progress(LogProtoServer *s)
+{
+  LogProtoAutoServer *self = (LogProtoAutoServer *) s;
+
+  if (self->proto_impl)
+    return log_proto_server_handshake_in_progress(self->proto_impl);
+
+  /* as long as the auto detection is not yet finished we are in handshake mode */
+  return TRUE;
+}
+
+static LogProtoStatus
+log_proto_auto_handshake(LogProtoServer *s)
+{
+  LogProtoAutoServer *self = (LogProtoAutoServer *) s;
+  /* allow the impl to do its handshake */
+  if (self->proto_impl)
+    return log_proto_server_handshake(self->proto_impl);
+
+  gchar detect_buffer[1];
+  gint rc;
+
+  rc = log_transport_read_ahead(self->super.transport, detect_buffer, sizeof(detect_buffer));
+  if (rc == 0)
+    return LPS_EOF;
+
+  if (rc != sizeof(detect_buffer))
```

Comment:
I think we should not return `LPS_ERROR` in short read cases.
Even with the current "we just need 1 byte" implementation, it would be good if we could handle EAGAIN peacefully.

Reply:
I have added support for this into the read_ahead implementation.

With that we could do a more elaborate testing of the initial bytes.